Fix redis_check_rdb() hang when rdb is FIFO

joehu21 opened the pull request:
When loading RDB over the named piped, redis_check_rdb() is hung at fopen, because fopen blocks until another process opens the FIFO for writing. The fix is to check if RDB is FIFO. If yes, return an error.

```
Release notes
Fix a bug where a failed RDB load might hang if loading from a FIFO instead of a file it wasn't open for write.
```

QuChen88 commented:
This change looks good to me.

madolson commented on src/redis-check-rdb.c:
```diff
@@ -199,6 +205,11 @@ int redis_check_rdb(char *rdbfilename, FILE *fp) {
     struct stat sb;
 
+    if (is_fifo(rdbfilename)) {
+        // Cannot check RDB over named pipe because fopen blocks until another process opens the FIFO for writing.
```
```suggestion
        /* Cannot check RDB over named pipe because fopen blocks until another process opens the FIFO for writing. */
```

madolson commented on src/redis-check-rdb.c:
```diff
@@ -186,6 +186,12 @@ void rdbCheckSetupSignals(void) {
     sigaction(SIGABRT, &act, NULL);
 }
 
+static int is_fifo(char *filename) {
```
```suggestion
static int isFifo(char *filename) {
```

madolson commented on src/redis-check-rdb.c:
```diff
@@ -199,6 +205,11 @@ int redis_check_rdb(char *rdbfilename, FILE *fp) {
     static rio rdb; /* Pointed by global struct riostate. */
     struct stat sb;
 
+    if (is_fifo(rdbfilename)) {
```
```suggestion
    if (isFifo(rdbfilename)) {
```

oranagra commented:
In which scenario do we run this code on a pipe?
If ohe doesn't exist in the OSS redis, then it shouldn't go on the release notes

oranagra commented:
Could this error be related to this change? 
32bit check-aof
https://github.com/redis/redis/actions/runs/4704875449/jobs/8344887796#step:6:4722

joehu21 commented:
@oranagra I encountered the "hang issue" in our Redis fork that loads RDB stream over a named pipe. I think the code of redis_check_rdb() is only intended for checking RDB disk file, not FIFO. Therefore, I considered it as a bug that the function does not verify the file is not FIFO. Note that other forks or applications may do the same thing - loading RDB over a pipe. I'm also okay with considering it as an enhancement rather than a bug.

The 32bit check-aof failure is likely related to this change. Do you know how to build a 32bit binary on a 64bit system? I tried "make 32bit" but got build errors.

oranagra commented:
If it's unreachable then it's not a bug and should not be mentioned in the release notes.
I'm all in favor of improving the code base so that it better serves forks and is better prepared to future changes in redis (depending of the complexity of the improvement of course).
I.e. I'm not against this change, just want it to be labeled correctly.

Regarding the 32bit, you can. Check the Readme, and also the github actions code (in the `.github` folder)

joehu21 commented:
@oranagra Built 32bit on ubuntu and tested aof.tcl. Did not see the error:
```
ubuntu@ip-172-31-85-103:~/github/redis$ ./src/redis-server --version
Redis server v=255.255.255 sha=d5d56d0d:0 malloc=jemalloc-5.2.1 bits=32 build=4485e9c7997cff7c

ubuntu@ip-172-31-85-103:~/github/redis$ ./runtest --single integration/aof
Cleanup: may take some time... OK
Starting test server at port 21079
[ready]: 10572
Testing integration/aof
[ready]: 10573
[ready]: 10575
[ready]: 10576
[ready]: 10574
[ready]: 10577
[ready]: 10578
[ready]: 10579
[ready]: 10580
[ready]: 10581
[ready]: 10582
[ready]: 10583
[ready]: 10584
[ready]: 10585
[ready]: 10586
[ready]: 10587
[ok]: Unfinished MULTI: Server should start if load-truncated is yes (2 ms)
[ok]: Short read: Server should start if load-truncated is yes (2 ms)
[ok]: Truncated AOF loaded: we expect foo to be equal to 5 (1 ms)
[ok]: Append a new command after loading an incomplete AOF (0 ms)
[ok]: Short read + command: Server should start (2 ms)
[ok]: Truncated AOF loaded: we expect foo to be equal to 6 now (1 ms)
[ok]: Bad format: Server should have logged an error (3 ms)
[ok]: Unfinished MULTI: Server should have logged an error (1 ms)
[ok]: Short read: Server should have logged an error (1 ms)
[ok]: Short read: Utility should confirm the AOF is not valid (4 ms)
[ok]: Short read: Utility should show the abnormal line num in AOF (5 ms)
[ok]: Short read: Utility should be able to fix the AOF (5 ms)
[ok]: Fixed AOF: Server should have been started (2 ms)
[ok]: Fixed AOF: Keyspace should contain values that were parseable (1 ms)
[ok]: AOF+SPOP: Server should have been started (2 ms)
[ok]: AOF+SPOP: Set should have 1 member (0 ms)
[ok]: AOF+SPOP: Server should have been started (2 ms)
[ok]: AOF+SPOP: Set should have 1 member (0 ms)
[ok]: AOF+EXPIRE: Server should have been started (2 ms)
[ok]: AOF+EXPIRE: List should be empty (1 ms)
[ok]: Redis should not try to convert DEL into EXPIREAT for EXPIRE -1 (1 ms)
[ok]: AOF fsync always barrier issue (5400 ms)
[ok]: GETEX should not append to AOF (0 ms)
[ok]: Unknown command: Server should have logged an error (2 ms)
[ok]: AOF+LMPOP/BLMPOP: pop elements from the list (102 ms)
[ok]: AOF+LMPOP/BLMPOP: after pop elements from the list (1 ms)
[ok]: AOF+ZMPOP/BZMPOP: pop elements from the zset (102 ms)
[ok]: AOF+ZMPOP/BZMPOP: after pop elements from the zset (1 ms)
[ok]: Generate timestamp annotations in AOF (369 ms)
[ok]: Successfully load AOF which has timestamp annotations inside (0 ms)
[ok]: Truncate AOF to specific timestamp (721 ms)
[ok]: EVAL timeout with slow verbatim Lua script from AOF (3410 ms)
[ok]: EVAL can process writes from AOF in read-only replicas (238 ms)
[ok]: Test redis-check-aof for old style resp AOF (5 ms)
[ok]: Test redis-check-aof for old style rdb-preamble AOF (7 ms)
[ok]: Test redis-check-aof for Multi Part AOF with resp AOF base (5 ms)
[ok]: Test redis-check-aof for Multi Part AOF with rdb-preamble AOF base (9 ms)
[ok]: Test redis-check-aof only truncates the last file for Multi Part AOF in fix mode (10 ms)
[ok]: Test redis-check-aof only truncates the last file for Multi Part AOF in truncate-to-timestamp mode (5 ms)
[ok]: FLUSHDB / FLUSHALL should persist in AOF (70 ms)
[1/1 done]: integration/aof (15 seconds)

                   The End

Execution time of different units:
  15 seconds - integration/aof

\o/ All tests passed without errors!

Cleanup: may take some time... OK
```

oranagra commented:
Edited the top comment, replaced the release notes section with a clarification.

regarding the test failure: i see two similar failures in again nightly, this time not in 32bit, instead it's in alpine and and valgrind with libc malloc.
https://github.com/redis/redis/actions/runs/4710428958/jobs/8354057551
it's the exact same failure as the one we had before in 32bit:
```
*** [err]: Test redis-check-aof for Multi Part AOF with rdb-preamble AOF base in tests/integration/aof.tcl
Expected 'Start checking Multi Part AOF
Start to check BASE AOF (RDB format).
[offset 0] Checking RDB file appendonly.aof.1.base.aof
RDB preamble of AOF file is not sane, aborting.
child process exited abnormally' to match '*Start checking Multi Part AOF*Start to check BASE AOF (RDB format)*DB preamble is OK, proceeding with AOF tail*BASE AOF*is valid*Start to check INCR files*INCR AOF*is valid*All AOF files and manifest are valid*' (context: type eval line 17 cmd {assert_match "*Start checking Multi Part AOF*Start to check BASE AOF (RDB format)*DB preamble is OK, proceeding with AOF tail*BASE AOF*is valid*Start to check INCR files*INCR AOF*is valid*All AOF files and manifest are valid*" $result} proc ::test) 
```
what we see is that for some reason the `redis_check_rdb_main` function returns C_ERR without printing anything (which is what the code in this PR does).
The fix in this PR seems rather safe, but it's a strange coincidence that these errors started after it was merged.
please try to take a deeper look.

joehu21 commented:
@oranagra First of all, the error is very likely caused by my commit. Second, I'm not able to reproduce the error using "./runtest --single integration/aof". Can you give me a pointer to how to run daily.yml (as it does in github actions)? I don't see any instruction in the README. Third, while I'm investigating the root cause, we can revert the commit.

joehu21 commented:
@oranagra  I think I've found the root cause. Submitted fix https://github.com/redis/redis/pull/12051. Note that since I have no luck reproducing the test failure, this is a fix in theory.
